core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -806,15 +808,20 @@ private[spark] object Utils extends Logging {
         // It's more proper to pick ip address following system output order.
         val activeNetworkIFs = NetworkInterface.getNetworkInterfaces.toList
         val reOrderedNetworkIFs = if (isWindows) activeNetworkIFs else activeNetworkIFs.reverse
+
         for (ni <- reOrderedNetworkIFs) {
-          for (addr <- ni.getInetAddresses if !addr.isLinkLocalAddress &&
-               !addr.isLoopbackAddress && addr.isInstanceOf[Inet4Address]) {
+          val addresses = ni.getInetAddresses.toList
+            .filter { addr => !addr.isLinkLocalAddress && !addr.isLoopbackAddress }
+          if ( !addresses.isEmpty ) {
+            val addr = addresses.find( _.isInstanceOf[Inet4Address] ).getOrElse(addresses.head)
+            // because of Inet6Address.toHostName may add interface at the end if it knows about it
+            val strippedAddress = InetAddress.getByAddress(addr.getAddress)
```

Comment:
It's all about finding our external IP address: old logic was to use external ipv4 || 127.0.0.1. 
That transformes it to ipv4 || ipv6 || 127.0.0.1

Whenewer we don't have any ipv4 it's better to use ipv6 than 127.0.0.1